csharp/src/Google.Protobuf/AsyncMessageParser.cs

```diff
@@ -0,0 +1,219 @@
+    /// </p>
+    /// <p>
+    /// Most applications will never need to create their own instances of this type;
+    /// instead, use the static <c>Parser</c> property of a generated message type to obtain a
+    /// parser for that type.
+    /// </p>
+    /// </remarks>
+    /// <typeparam name="T">The type of message to be parsed.</typeparam>
+    public sealed class AsyncMessageParser<T> : MessageParser<T>, IAsyncMessageParser where T : IAsyncMessage<T>
+    {
+        /// <summary>
+        /// Creates a new parser.
+        /// </summary>
+        /// <remarks>
+        /// The factory method is effectively an optimization over using a generic constraint
+        /// to require a parameterless constructor: delegates are significantly faster to execute.
+        /// </remarks>
+        /// <param name="factory">Function to invoke when a new, empty message is required.</param>
+        public AsyncMessageParser(Func<T> factory) : base(factory) { }
+
+        /// <summary>
+        /// Parses a message from the given coded input stream.
+        /// </summary>
+        /// <param name="input">The stream to parse.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>The parsed message.</returns>
+        public async Task<T> ParseFromAsync(CodedInputStream input, CancellationToken cancellationToken)
+        {
+            T message = CreateTemplate();
+            await message.MergeFromAsync(input, cancellationToken).ConfigureAwait(false);
+            return message;
+        }
+
+        /// <summary>
+        /// Parses a message from the given stream.
+        /// </summary>
+        /// <param name="input">The stream to parse.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>The parsed message.</returns>
+        public async Task<T> ParseFromAsync(Stream input, CancellationToken cancellationToken)
+        {
+            T message = CreateTemplate();
+            await message.MergeFromAsync(input, cancellationToken).ConfigureAwait(false);
+            return message;
+        }
+
+        /// <summary>
+        /// Parses a length-delimited message from the given stream.
+        /// </summary>
+        /// <remarks>
+        /// The stream is expected to contain a length and then the data. Only the amount of data
+        /// specified by the length will be consumed.
+        /// </remarks>
+        /// <param name="input">The stream to parse.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>The parsed message.</returns>
+        public async Task<T> ParseDelimitedFromAsync(Stream input, CancellationToken cancellationToken)
+        {
+            T message = CreateTemplate();
+            await message.MergeFromAsync(input, cancellationToken).ConfigureAwait(false);
```

Comment:
Shouldn't this use MergeDelimitedFromAsync instead?

Reply:
Yup it should. Thanks for finding this out!